web/turkey/survey/views.py

```diff
@@ -1,106 +1,215 @@
 from django.apps import apps
 from django.contrib.auth.mixins import LoginRequiredMixin
+from django.contrib.staticfiles.templatetags.staticfiles import static
 from django.core.paginator import Paginator
+from django.core.urlresolvers import reverse
 from django.db import transaction
-from django.http import Http404, StreamingHttpResponse, HttpResponse
+from django.http import Http404, StreamingHttpResponse
 from django.template.response import TemplateResponse
-from django.utils.decorators import method_decorator
+from django.utils.safestring import mark_safe
 from django.utils.translation import ugettext_lazy as _
-from django.views.decorators.csrf import ensure_csrf_cookie
 from django.views.generic import View
 from rest_framework import status
+from rest_framework.authentication import BaseAuthentication
 from rest_framework.exceptions import ValidationError
+from rest_framework.permissions import AllowAny
 from rest_framework.views import APIView
 from rest_framework.response import Response
 
 from .auditors import NAME_TO_AUDITOR
 from .steps import NAME_TO_STEP
 from .models import Task, TaskInteraction
 
-
 # TODO: Authentication
-from survey.renderers import XMLBodyRenderer
+from .renderers import XMLBodyRenderer
+
+
+# TODO: Implement token authentication. So when the interaction is created, make a token to identify user...
+class NonAuthentication(BaseAuthentication):
+    def authenticate(self, request):
+        pass
 
 
 class RecordSubmission(APIView):
```

Comment:
Lots of refactoring of this class to account for
1) The lack of an explicit 'completed' flag, necessitating queries
2) The splitting of auditor and step submission into two different classes that now inherit from this (so this is now a parent class)
3) Instead of relying on the user provided object to tell us what classes to look up and save to, we should know what classes we want and be trying to look those up (in other words I changed a flaw that allowed users to modify our code flow and cause us to call their input valid when it was missing data)
4) Lock on task_interaction using select_for_update while we process a submission so if a submission fires two times there is no race. When we finish processing one, we will begin trying to process the other, but we will see all the updates and bail.
5) Finally dealing with the edge case of a logged in admin
6) Formalizing permissions